flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/sink/RowDataTaskWriterFactory.java

```diff
@@ -69,8 +71,16 @@ public RowDataTaskWriterFactory(Table table,
 
     if (equalityFieldIds == null || equalityFieldIds.isEmpty()) {
       this.appenderFactory = new FlinkAppenderFactory(schema, flinkSchema, table.properties(), spec);
+    } else if (upsert) {
+      // In upsert mode, only the new row is emitted using INSERT row kind. Therefore, the deletion schema
+      // must be only the equality fields for that the writer can use the new row to issue a delete by key.
```

Comment:
Minor: The second sentence doesn't naturally follow from the first. The link is that any column of the inserted row may differ from the deleted row, and the delete file must contain values that are correct for the deleted row. Therefore, only write the equality delete fields.

Reply:
Updated